File: app/android/app/src/main/java/com/jun/weather/ui/component/WeeklyItemLayout.java
package com.jun.weather.ui.component;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;

import androidx.constraintlayout.widget.ConstraintLayout;

import com.jun.weather.R;

public class WeeklyItemLayout extends ConstraintLayout {
    private WeeklyItemView[] weeklyItemViews;

    public WeeklyItemLayout(Context context) {
        super(context);
        initView(context);
    }

    public WeeklyItemLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(context);
        getAttrs(attrs);
    }

    public WeeklyItemLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context);
        getAttrs(attrs, defStyleAttr);
    }

    private void initView(Context context) {
        View view = LayoutInflater.from(context).inflate(R.layout.component_weekly_layout, this, false);
        addView(view);

        weeklyItemViews = new WeeklyItemView[8];
        weeklyItemViews[1] = findViewById(R.id.weekly_item_1);
        weeklyItemViews[2] = findViewById(R.id.weekly_item_2);
        weeklyItemViews[3] = findViewById(R.id.weekly_item_3);
        weeklyItemViews[4] = findViewById(R.id.weekly_item_4);
        weeklyItemViews[5] = findViewById(R.id.weekly_item_5);
        weeklyItemViews[6] = findViewById(R.id.weekly_item_6);
        weeklyItemViews[7] = findViewById(R.id.weekly_item_7);
    }

    private void getAttrs(AttributeSet attrs) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.WeeklyItemLayout);
        setAttrs(typedArray);
    }

    private void getAttrs(AttributeSet attrs, int defStyleAttr) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.WeeklyItemLayout, defStyleAttr, 0);
        setAttrs(typedArray);
    }

    private void setAttrs(TypedArray typedArray) {
        weeklyItemViews[1].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item1_visibility, View.VISIBLE));
        weeklyItemViews[2].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item2_visibility, View.VISIBLE));
        weeklyItemViews[3].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item3_visibility, View.VISIBLE));
        weeklyItemViews[4].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item4_visibility, View.VISIBLE));
        weeklyItemViews[5].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item5_visibility, View.VISIBLE));
        weeklyItemViews[6].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item6_visibility, View.VISIBLE));
        weeklyItemViews[7].setVisibility(typedArray.getInt(R.styleable.WeeklyItemLayout_item7_visibility, View.VISIBLE));

        typedArray.recycle();
    }

    public void setWeeklyItemVisibility(int idx, int val) {
        try {
            if(val != View.VISIBLE && val != View.GONE && val != View.INVISIBLE) {
                throw new Exception("value is not visibility constant.");
            }

            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setVisibility(val);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setDateText(int idx, String date) {
        try {
            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setDateText(date);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setSkyImage(int idx, int dayResourceId, int nightResourceId) {
        try {
            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setSkyDayImage(dayResourceId);
            weeklyItemViews[idx].setSkyNightImage(nightResourceId);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setLowTemp(int idx, int i) {
        try {
            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setLowTemp(i);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setLowTemp(int idx, String val) {
        try {
            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setLowTemp(val);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setHighTemp(int idx, int i) {
        try {
            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setHighTemp(i);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setHighTemp(int idx, String val) {
        try {
            if(idx > 7 || idx < 1) {
                throw new Exception("item index exceed.");
            }

            weeklyItemViews[idx].setHighTemp(val);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/android/app/src/main/java/com/jun/weather/ui/component/WeeklyItemLayout.kt
package com.jun.weather.ui.component

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.view.LayoutInflater
import androidx.constraintlayout.widget.ConstraintLayout
import com.jun.weather.R
import com.jun.weather.databinding.ComponentWeeklyLayoutBinding
import com.jun.weather.databinding.FragmentWeekWeatherBinding

class WeeklyItemLayout : ConstraintLayout {
    private lateinit var weeklyItemViews: Array<WeeklyItemView?>
    private lateinit var binding: ComponentWeeklyLayoutBinding

    constructor(context: Context) : super(context) {
        initView(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initView(context)
        getAttrs(attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initView(context)
        getAttrs(attrs, defStyleAttr)
    }

    private fun initView(context: Context) {
        val view = LayoutInflater.from(context).inflate(R.layout.component_weekly_layout, this, false)
        binding = ComponentWeeklyLayoutBinding.bind(view)
        addView(view)

        weeklyItemViews = arrayOfNulls(8)
        weeklyItemViews[1] = binding.weeklyItem1
        weeklyItemViews[2] = binding.weeklyItem2
        weeklyItemViews[3] = binding.weeklyItem3
        weeklyItemViews[4] = binding.weeklyItem4
        weeklyItemViews[5] = binding.weeklyItem5
        weeklyItemViews[6] = binding.weeklyItem6
        weeklyItemViews[7] = binding.weeklyItem7
    }

    private fun getAttrs(attrs: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.WeeklyItemLayout)
        setAttrs(typedArray)
    }

    private fun getAttrs(attrs: AttributeSet?, defStyleAttr: Int) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.WeeklyItemLayout, defStyleAttr, 0)
        setAttrs(typedArray)
    }

    private fun setAttrs(typedArray: TypedArray) {
        weeklyItemViews[1]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item1_visibility, VISIBLE)
        weeklyItemViews[2]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item2_visibility, VISIBLE)
        weeklyItemViews[3]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item3_visibility, VISIBLE)
        weeklyItemViews[4]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item4_visibility, VISIBLE)
        weeklyItemViews[5]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item5_visibility, VISIBLE)
        weeklyItemViews[6]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item6_visibility, VISIBLE)
        weeklyItemViews[7]!!.visibility = typedArray.getInt(R.styleable.WeeklyItemLayout_item7_visibility, VISIBLE)
        typedArray.recycle()
    }

    fun setWeeklyItemVisibility(idx: Int, `val`: Int) {
        try {
            if (`val` != VISIBLE && `val` != GONE && `val` != INVISIBLE) {
                throw Exception("value is not visibility constant.")
            }
            if (idx > 7 || idx < 1) {
                throw Exception("item index exceed.")
            }
            weeklyItemViews[idx]!!.visibility = `val`
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun setDateText(idx: Int, date: String?) {
        try {
            if (idx > 7 || idx < 1) {
                throw Exception("item index exceed.")
            }
            weeklyItemViews[idx]!!.setDateText(date)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun setSkyImage(idx: Int, dayResourceId: Int, nightResourceId: Int) {
        try {
            if (idx > 7 || idx < 1) {
                throw Exception("item index exceed.")
            }
            weeklyItemViews[idx]!!.setSkyDayImage(dayResourceId)
            weeklyItemViews[idx]!!.setSkyNightImage(nightResourceId)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun setTemp(idx: Int, low: String, high: String) {
        try {
            if (idx > 7 || idx < 1) {
                throw Exception("item index exceed.")
            }
            weeklyItemViews[idx]!!.setTemp(low, high)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}